Tests/DependencyInjection/Fixtures/Bundles/RepositoryServiceBundle/Repository/TestCustomRepoRepository.php

```diff
@@ -0,0 +1,27 @@
+<?php
+
+/*
+ * This file is part of the Doctrine Bundle
+ *
+ * The code was originally distributed inside the Symfony framework.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ * (c) Doctrine Project, Benjamin Eberlei <kontakt@beberlei.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Fixtures\Bundles\RepositoryServiceBundle\Repository;
+
+use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
+use Fixtures\Bundles\RepositoryServiceBundle\Entity\TestCustomRepoEntity;
+use Symfony\Bridge\Doctrine\RegistryInterface;
+
+class TestCustomRepoRepository extends ServiceEntityRepository
+{
+    public function __construct(RegistryInterface $registry)
```

Comment:
Having to override the constructor is going to feel weird. People are likely to forget to override it or to change the signature. One option to avoid this is to loop through the metadata array from the entity manager and look for an entity that has our class as custom repository. It would be nice if we could even move that to the service itself to avoid looking up that information on every instantiation, but i believe changing this will improve DX.

Reply:
The need to override the constructor is indeed the weirdest part :/ (but you only need to do it if you need to add more constructor args, otherwise the default constructor is fine). So, I’m interested in your idea, but I don’t understand yet. Where would we loop over the class metadata? And how would that help? Doesn’t the user need to (somehow) do some work/chores in their service to set the protected properties that EntityRepository needs internally?

Reply:
I thought that the constructor override was necessary to get the new repositories to play nice with service auto-registration. Since there will most likely be a repository for a single entity (unlike the default `EntityRepository`), it would make sense trying to deduct the entity class name from the entity manager in `ServiceEntityRepository` if it isn't given:
- If entity class name is missing, fetch a list of all metadata from the document managers (yes, plural unfortunately).
- Loop over metadata, looking for an entity with a `customRepositoryClass` of `get_class($this)`.
- If a single entity is found, use it. If more/less than one entity is found, throw an exception.